File: parser/src/main/java/es/ulpgc/deserializers/CodeUploadEventDeserializer.java
package es.ulpgc.deserializers;

import com.google.gson.*;
import es.ulpgc.events.CodeUploadEvent;

import java.lang.reflect.Type;
import java.time.Instant;

public class CodeUploadEventDeserializer {
    public CodeUploadEvent deserialize(String event) {
        Gson gson = new GsonBuilder().registerTypeAdapter(CodeUploadEvent.class, new CodeUploadTypeAdapter()).create();
        return gson.fromJson(event, CodeUploadEvent.class);
    }

    private static class CodeUploadTypeAdapter implements JsonDeserializer<CodeUploadEvent> {
        @Override
        public CodeUploadEvent deserialize(JsonElement json, Type type, JsonDeserializationContext content) throws JsonParseException {
            JsonObject jsonObject = json.getAsJsonObject();
            return new CodeUploadEvent(
                    Instant.parse(jsonObject.get("time").getAsString()),
                    jsonObject.get("source").getAsString(),
                    jsonObject.get("detail").getAsJsonObject().get("filename").getAsString()
            );
        }
    }
}

File: tokenizer/code_upload_event_deserializer.py
from datetime import datetime

from events.code_upload_event import CodeUploadEvent


class AwsCodeUploadEventDeserializer:
    def deserialize(self, _object: str):
        ts = self.__to_timestamp(_object["time"])
        source = _object["source"]
        account = _object["account"]
        region = _object["region"]
        filename = _object["detail"]["Filename"]
        return CodeUploadEvent(ts, source, account, region, filename)

    def __to_timestamp(self, time):
        return datetime.strptime(time, "%Y-%m-%dT%H:%M:%SZ").timestamp()
